From: Hannes Reinecke <hare@suse.de>
Date: Wed, 4 Apr 2018 14:12:32 +0200
Subject: [PATCH] dm: Always copy cmd_flags when cloning a request
Patch-Mainline: never, SLE12-SP3 specific
References: bsc#1088087

When cloning a request we should be copying over the 'cmd_flags'
from the original request; we need to modify the flags ourselves
so there is not much sense having the block layer doing that.
This mimics upstream behaviour.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/md/dm-mpath.c | 1 +
 drivers/md/dm-rq.c    | 1 +
 2 files changed, 2 insertions(+)

diff --git a/drivers/md/dm-mpath.c b/drivers/md/dm-mpath.c
index de5705a0d539..7896e11805f1 100644
--- a/drivers/md/dm-mpath.c
+++ b/drivers/md/dm-mpath.c
@@ -583,6 +583,7 @@ static int __multipath_map(struct dm_target *ti, struct request *clone,
 		}
 		clone->bio = clone->biotail = NULL;
 		clone->rq_disk = bdev->bd_disk;
+		clone->cmd_flags = rq->cmd_flags | REQ_NOMERGE;
 		clone->cmd_flags |= REQ_FAILFAST_TRANSPORT;
 		*__clone = clone;
 	}
diff --git a/drivers/md/dm-rq.c b/drivers/md/dm-rq.c
index a0701a7fd016..d1a93c0e4b9f 100644
--- a/drivers/md/dm-rq.c
+++ b/drivers/md/dm-rq.c
@@ -498,6 +498,7 @@ static int setup_clone(struct request *clone, struct request *rq,
 	if (r)
 		return r;
 
+	clone->cmd_flags = rq->cmd_flags | REQ_NOMERGE;
 	clone->end_io = end_clone_request;
 	clone->end_io_data = tio;
 
-- 
2.12.3

